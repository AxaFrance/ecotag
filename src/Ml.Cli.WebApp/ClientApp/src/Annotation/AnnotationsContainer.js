import React, {useEffect, useState} from "react";
import AnnotationSwitch from "./AnnotationSwitch";
import AnnotationsToolbar from "./AnnotationsToolbar";
import './AnnotationsContainer.scss';
import {useHistory} from "react-router";
import {useMutation} from "react-query";
import {fetchGetData, fetchPostJson, utf8_to_b64} from "../FetchHelper";
import {toast} from "react-toastify";

const selectItemById = (annotationState, id) => {
    if(id === "end")
        return null;
    if(id === "empty")
        return null;
    return annotationState.items.find(x => x.id === id);
};


const fetchImages = async data => {
    if (data.status === 200) {
        const hardDriveLocations = await data.json();
        return hardDriveLocations.map(element => {
            return `api/files/${utf8_to_b64(element)}`;
        });
    } else {
        return [];
    }
};

const getImages = (fetchFunction) => async (item) => {
    const params = "fileName=" + item.fileName + "&stringsMatcher="+ item.frontDefaultStringsMatcher + "&directory=" + item.imageDirectory
    const fetchResult = await fetchGetData(fetchFunction)("api/datasets/"+utf8_to_b64(params));
    return fetchImages(fetchResult);
};

const sendConfirmationMessage = (isSuccess) => {
    const message = isSuccess ? "Annotation saved" : "Impossible to save annotation";
    const type = isSuccess ? "success" : "error";
    toast(message, {
        position: "top-left",
        autoClose: 800,
        hideProgressBar: true,
        closeOnClick: true,
        pauseOnHover: false,
        draggable: true,
        progress: undefined,
        type: type
    });
}

const AnnotationsContainer = ({state, id, url, dataset, fetchFunction}) => {
    const history = useHistory();
    
    const item = selectItemById(state, id);
    const itemNumber= state.items.indexOf(item);
    
    const onSubmit = async (annotation) => {
        const annotationObject = {
            datasetLocation: state.datasetLocation,
            annotationType: state.annotationType,
            fileName: item.fileName,
            annotation: annotation
        };
        await mutationDataset.mutateAsync(annotationObject, {
            onSuccess: () => {
                sendConfirmationMessage(true);
                onNext();
            },
            onError: () => sendConfirmationMessage(false)
        });
    }
    
    const onPrevious = () => {
        if(id === "end"){
            history.push(`${url}/${dataset}/${state.items[state.items.length - 1].id}`);
        } else{
            history.push(`${url}/${dataset}/${state.items[itemNumber - 1].id}`);
        }
    };
    
    const onNext = () => {
        if((itemNumber + 1) < state.items.length){
            history.push(`${url}/${dataset}/${state.items[itemNumber + 1].id}`);
        } else {
            history.push(`${url}/${dataset}/end`);
        }
    };

    const isPreviousDisabled = itemNumber === 0;
    const isNextDisabled = item == null;

    const [localState, setState] = useState({
        filePrimaryUrl: "",
    });
    console.log(item);

    const mutationDataset = useMutation(newData => fetchPostJson(fetchFunction)("/api/annotations", newData));

    const getUrls = async () => {
        const newUrls = await getImages(fetchFunction)(item);
        const newFileUrl = newUrls != null ? newUrls[0] : "";
        return {fileUrls: newUrls, filePrimaryUrl: newFileUrl};
    };

    const getBoundingBoxes = () => {
        let boundingBoxes;
        try{
            const annotationsArray = JSON.parse(item.annotations);
            const lastAnnotation = annotationsArray[annotationsArray.length - 1];
            const labels = lastAnnotation.labels;
            boundingBoxes = labels.boundingBoxes;
        }
        catch (ex){
            boundingBoxes = [];
        }
        return boundingBoxes;
    }

    useEffect(() => {
        setState({
            filePrimaryUrl: "",
        })
        if(!item || !item.imageDirectory){
            return;
        }
        getUrls().then(urls => {
                setState({
                    filePrimaryUrl: urls.filePrimaryUrl,
            })
        });
    }, [id]);

    if(state.items.length <= 0){
        return <h2 className="error-message">The annotation file is empty.</h2>
    }

    return <>
        <AnnotationsToolbar
            onPrevious={onPrevious}
            isPreviousDisabled={isPreviousDisabled}
            onNext={onNext}
            text={item? item.fileName: ""}
            isNextDisabled={isNextDisabled}
        />
        {isNextDisabled ? (
            <h3 className="annotation__end-message">Thank you, all files from this dataset have been annotated.</h3>
        ) : (
            <AnnotationSwitch
                url={localState.filePrimaryUrl}
                annotationType={state.annotationType}
                labels={state.configuration}
                expectedOutput={getBoundingBoxes()}
                onSubmit={onSubmit}
            />
        )}
    </>;
};

export default React.memo(AnnotationsContainer);
