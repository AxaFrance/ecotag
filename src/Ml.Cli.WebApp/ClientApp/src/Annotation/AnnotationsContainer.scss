.end-message{
  color: green;
  text-align: center;
}