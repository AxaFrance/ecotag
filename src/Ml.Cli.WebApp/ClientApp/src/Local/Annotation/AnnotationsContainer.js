import React, {useEffect, useState} from "react";

import {useHistory} from "react-router";
import {useMutation} from "react-query";
import {fetchGetData, fetchPostJson, utf8_to_b64} from "../FetchHelper";
import {toast} from "react-toastify";
import Annotations from "../../Toolkit/Annotations/Annotations";

const selectItemById = (annotationState, id) => {
    if(id === "end")
        return null;
    if(id === "empty")
        return null;
    return annotationState.items.find(x => x.id === id);
};


const fetchImages = async data => {
    if (data.status === 200) {
        const hardDriveLocations = await data.json();
        return hardDriveLocations.map(element => {
            return `api/local/files/${utf8_to_b64(element)}`;
        });
    } else {
        return [];
    }
};

const getImages = (fetch) => async (item) => {
    const params = "fileName=" + item.fileName + "&stringsMatcher="+ item.frontDefaultStringsMatcher + "&directory=" + item.imageDirectory
    const fetchResult = await fetchGetData(fetch)("api/local/datasets/"+utf8_to_b64(params));
    return fetchImages(fetchResult);
};

const sendConfirmationMessage = (isSuccess) => {
    const message = isSuccess ? "Annotation saved" : "Impossible to save annotation";
    const type = isSuccess ? "success" : "error";
    toast(message, {
        position: "top-left",
        autoClose: 800,
        hideProgressBar: true,
        closeOnClick: true,
        pauseOnHover: false,
        draggable: true,
        progress: undefined,
        type: type
    });
}


const AnnotationsContainer = ({state, id, url, dataset, fetch}) => {
    const history = useHistory();
    
    const item = selectItemById(state, id);
    const itemNumber= state.items.indexOf(item);
    
    const onSubmit = async (annotation) => {
        const annotationObject = {
            datasetLocation: state.datasetLocation,
            annotationType: state.annotationType,
            fileName: item.fileName,
            annotation: annotation
        };
        await mutationDataset.mutateAsync(annotationObject, {
            onSuccess: () => {
                sendConfirmationMessage(true);
                onNext();
            },
            onError: () => sendConfirmationMessage(false)
        });
    }
    
    const onPrevious = () => {
        if(id === "end"){
            history.push(`${url}/${dataset}/${state.items[state.items.length - 1].id}`);
        } else{
            history.push(`${url}/${dataset}/${state.items[itemNumber - 1].id}`);
        }
    };
    
    const onNext = () => {
        if((itemNumber + 1) < state.items.length){
            history.push(`${url}/${dataset}/${state.items[itemNumber + 1].id}`);
        } else {
            history.push(`${url}/${dataset}/end`);
        }
    };

    const isPreviousDisabled = itemNumber === 0;
    const isNextDisabled = item == null;

    const [localState, setState] = useState({
        filePrimaryUrl: "",
    });

    const mutationDataset = useMutation(newData => fetchPostJson(fetch)("/api/local/annotations", newData));

    const getUrls = async () => {
        const newUrls = await getImages(fetch)(item);
        const newFileUrl = newUrls != null ? newUrls[0] : "";
        return {fileUrls: newUrls, filePrimaryUrl: newFileUrl};
    };

    const getBoundingBoxes = () => {
        if(item && item.annotations){
            const annotationsArray = JSON.parse(item.annotations);
            const lastAnnotation = annotationsArray[annotationsArray.length - 1];
            return lastAnnotation.labels.boundingBoxes;
        }
        return null;
    }

    useEffect(() => {
        setState({
            filePrimaryUrl: "",
        })
        if(!item || !item.imageDirectory){
            return;
        }
        getUrls().then(urls => {
                setState({
                    filePrimaryUrl: urls.filePrimaryUrl,
            })
        });
    }, [id]);
    
    
    const toolbarText = item ? item.fileName: "";
    const annotationType = state.annotationType;
    const labels = state.configuration;
    const expectedOutput = getBoundingBoxes();

    return <Annotations
            onPrevious={onPrevious}
            isPreviousDisabled={isPreviousDisabled}
            onNext={onNext}
            toolbarText={toolbarText}
            isEmpty={state.items.length <= 0}
            isNextDisabled={isNextDisabled}
            url={localState.filePrimaryUrl}
            annotationType={annotationType}
            labels={labels}
            expectedOutput={expectedOutput}
            onSubmit={onSubmit}
        />;
};

export default React.memo(AnnotationsContainer);
