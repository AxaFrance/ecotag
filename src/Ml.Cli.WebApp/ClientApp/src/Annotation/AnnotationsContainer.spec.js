import AnnotationsContainer from "./AnnotationsContainer";
import React from "react";
import {fireEvent, render, waitFor} from "@testing-library/react";
import {QueryClient, QueryClientProvider} from "react-query";

const mockedFetchFunction = () => {};
const MonacoEditor = {};

const items = [
    {
        fileName: "first_file",
        id: 0,
        fileDirectory: "fileDirectory",
        imageDirectory: "imageDirectory",
        frontDefaultStringsMatcher: "",
        annotations: {}
    },
    {
        fileName: "second_file",
        id: 1,
        fileDirectory: "fileDirectory",
        imageDirectory: "imageDirectory",
        frontDefaultStringsMatcher: "",
        annotations: {}
    },
    {
        fileName: "third_file",
        id: 2,
        fileDirectory: "fileDirectory",
        imageDirectory: "imageDirectory",
        frontDefaultStringsMatcher: "",
        annotations: {}
    },
    {
        fileName: "fourth_file",
        id: 3,
        fileDirectory: "fileDirectory",
        imageDirectory: "imageDirectory",
        frontDefaultStringsMatcher: "",
        annotations: {}
    }
];

const state= {
    fileName: "dataset-NER.json",
    isFileInserted: true,
    items: items,
    datasetLocation: "datasetLocation",
    annotationType: "NamedEntityRecognition",
    configuration: [{id:"#008194",name:"Firstname",color:"#008194"},{id:"#00ffa2",name:"Lastname",color:"#00ffa2"}]
};

const client = new QueryClient();

describe("Should check AnnotationsContainer behaviour", () => {
    test("Should render properly and change items", async () => {
        
        const {container, asFragment, getByText} = render(<QueryClientProvider client={client}><AnnotationsContainer state={state} entryItem={items[0]} fetchFunction={mockedFetchFunction} MonacoEditor={MonacoEditor}/></QueryClientProvider>);
        
        await waitFor(() => expect(container.querySelector(".annotation__top-toolbar")).not.toBeNull());
        await waitFor(() => expect(container.querySelector(".table-result__header")).not.toBeNull());
        expect(asFragment()).toMatchSnapshot();
        
        const nextButton = getByText("Suivant");
        fireEvent.click(nextButton);
        
        await waitFor(() => expect(getByText(/second_file/i)).not.toBeNull());
        expect(asFragment()).toMatchSnapshot();
        
    });
});