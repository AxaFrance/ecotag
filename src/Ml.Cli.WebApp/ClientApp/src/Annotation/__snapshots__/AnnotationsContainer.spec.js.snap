// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should check AnnotationsContainer behaviour Should render properly and change items 1`] = `
<DocumentFragment>
  <div
    class="annotation__top-toolbar"
  >
    <button
      class="btn af-btn af-btn--hasiconLeft af-btn--disabled"
      disabled=""
      glyphicon="glyphicon glyphicon-arrowthin-left"
    >
      Précédent
    </button>
    <button
      class="btn af-btn af-btn--hasiconRight"
      glyphicon="glyphicon glyphicon-arrowthin-right"
    >
      Suivant
    </button>
  </div>
  <div
    class="table-result"
  >
    <div
      class="table-result__header"
    >
      <div>
        <p>
          File result: first_file
        </p>
      </div>
    </div>
    <div>
      Loading NamedEntity...
    </div>
  </div>
</DocumentFragment>
`;

exports[`Should check AnnotationsContainer behaviour Should render properly and change items 2`] = `
<DocumentFragment>
  <div
    class="annotation__top-toolbar"
  >
    <button
      class="btn af-btn af-btn--hasiconLeft"
      glyphicon="glyphicon glyphicon-arrowthin-left"
    >
      Précédent
    </button>
    <button
      class="btn af-btn af-btn--hasiconRight"
      glyphicon="glyphicon glyphicon-arrowthin-right"
    >
      Suivant
    </button>
  </div>
  <div
    class="table-result"
  >
    <div
      class="table-result__header"
    >
      <div>
        <p>
          File result: second_file
        </p>
      </div>
    </div>
    <div>
      <div
        class="annotationContainer"
      >
        <div
          class="sticky"
        >
          <div
            style="width: 100%;"
          >
            <div
              style="display: flex; width: 100%; color: rgb(255, 255, 255); padding: 15px 20px; font-size: 20px; text-align: center; box-sizing: border-box; font-weight: bold; font-family: \\"Roboto Condensed\\", \\"Arial Narrow\\", sans-serif; text-transform: uppercase; border-top-left-radius: 0; border-top-right-radius: 0;"
            >
              <div
                style="display: flex; flex-wrap: wrap;"
              >
                <span>
                  <label
                    for="Firstname"
                    style="margin: 5px 10px 5px 0px; border: 1px solid; cursor: pointer; font-size: 25px; padding: 1px 10px; position: relative; border-radius: 5px; color: rgb(75, 72, 72); background: rgb(0, 129, 148);"
                  >
                    Firstname
                  </label>
                </span>
                <span>
                  <label
                    for="Lastname"
                    style="margin: 5px 10px 5px 0px; border: 1px solid; cursor: pointer; font-size: 25px; padding: 1px 10px; position: relative; border-radius: 5px; color: rgb(75, 72, 72); background: rgb(0, 255, 162);"
                  >
                    Lastname
                  </label>
                </span>
              </div>
            </div>
          </div>
        </div>
        <div>
          Selected tag : Firstname
        </div>
        <div
          class="tokenAnnotation-container"
        >
          <div
            class="zoom-button-container"
          >
            <a
              class="btn af-btn--circle"
              href="#"
              role="button"
              tabindex="0"
            >
              <i
                class="glyphicon glyphicon-zoom-out"
              />
            </a>
            <a
              class="btn af-btn--circle"
              href="#"
              role="button"
              tabindex="0"
            >
              <i
                class="glyphicon glyphicon-zoom-in"
              />
            </a>
          </div>
          <div
            class="token-container"
            style="font-size: 14px;"
          >
            <span
              class="c0132"
              data-i="0"
            >
              Hello
            </span>
            <span
              class="c0132"
              data-i="1"
            >
              ,
            </span>
            <span
              class="c0132"
              data-i="2"
            >
              my
            </span>
            <span
              class="c0132"
              data-i="3"
            >
              name
            </span>
            <span
              class="c0132"
              data-i="4"
            >
              is
            </span>
            <span
              class="c0132"
              data-i="5"
            >
              Lilian
            </span>
            <span
              class="c0132"
              data-i="6"
            >
              Delouvy
            </span>
            <span
              class="c0132"
              data-i="7"
            >
              ,
            </span>
            <span
              class="c0132"
              data-i="8"
            >
              and
            </span>
            <span
              class="c0132"
              data-i="9"
            >
              this
            </span>
            <span
              class="c0132"
              data-i="10"
            >
              is
            </span>
            <span
              class="c0132"
              data-i="11"
            >
              some
            </span>
            <span
              class="c0132"
              data-i="12"
            >
              sample
            </span>
            <span
              class="c0132"
              data-i="13"
            >
              text
            </span>
            <span
              class="c0132"
              data-i="14"
            >
              .
            </span>
            <span
              class="c0132"
              data-i="15"
            >
              The
            </span>
            <span
              class="c0132"
              data-i="16"
            >
              NER
            </span>
            <span
              class="c0132"
              data-i="17"
            >
              is
            </span>
            <span
              class="c0132"
              data-i="18"
            >
              currently
            </span>
            <span
              class="c0132"
              data-i="19"
            >
              in
            </span>
            <span
              class="c0132"
              data-i="20"
            >
              a
            </span>
            <span
              class="c0132"
              data-i="21"
            >
              '
            </span>
            <span
              class="c0132"
              data-i="22"
            >
              Work
            </span>
            <span
              class="c0132"
              data-i="23"
            >
              In
            </span>
            <span
              class="c0132"
              data-i="24"
            >
              Progress
            </span>
            <span
              class="c0132"
              data-i="25"
            >
              '
            </span>
            <span
              class="c0132"
              data-i="26"
            >
              state
            </span>
            <span
              class="c0132"
              data-i="27"
            >
              .
            </span>
            <span
              class="c0132"
              data-i="28"
            >
              And
            </span>
            <span
              class="c0132"
              data-i="29"
            >
              it
            </span>
            <span
              class="c0132"
              data-i="30"
            >
              will
            </span>
            <span
              class="c0132"
              data-i="31"
            >
              be
            </span>
            <span
              class="c0132"
              data-i="32"
            >
              awesome
            </span>
            <span
              class="c0132"
              data-i="33"
            >
              .
            </span>
            <span
              class="c0132"
              data-i="34"
            >
              bla
            </span>
            <span
              class="c0132"
              data-i="35"
            >
              bla
            </span>
            <span
              class="c0132"
              data-i="36"
            >
              .
            </span>
            <span
              class="c0132"
              data-i="37"
            >
              bla
            </span>
            <span
              class="c0132"
              data-i="38"
            >
              bla
            </span>
            <span
              class="c0132"
              data-i="39"
            >
              .
            </span>
            <span
              class="c0132"
              data-i="40"
            >
              bla
            </span>
            <span
              class="c0132"
              data-i="41"
            >
              bla
            </span>
            <span
              class="c0132"
              data-i="42"
            >
              .
            </span>
            <span
              class="c0132"
              data-i="43"
            >
              bla
            </span>
            <span
              class="c0132"
              data-i="44"
            >
              bla
            </span>
            <span
              class="c0132"
              data-i="45"
            >
              .
            </span>
            <span
              class="c0132"
              data-i="46"
            >
              bla
            </span>
            <span
              class="c0132"
              data-i="47"
            >
              bla
            </span>
          </div>
        </div>
        <div
          class="annotationActionContainer"
        >
          <button
            class="buttonAnnotationAction"
          >
            Submit Annotation
          </button>
        </div>
        <pre>
          []
        </pre>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
